File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cpp
#include<vector>
#include<queue>
using namespace std;

int visit[104][104];
int dy[]={0,0,-1,1};
int dx[]={-1,1,0,0};

int solution(vector<vector<int> > maps)
{
    int answer = 0;
    
    queue<pair<int,int>>q;
    
    q.push({0,0}); // 시작 지점
    visit[0][0]=1;
    
    int maxCol = maps.size();
    int maxRow = maps[0].size();
    
    while(q.size()){
        
        pair<int,int>p = q.front();
        q.pop();
        int y=p.first;
        int x=p.second;
        
        for(int i=0;i<4;i++){
            int ny=y+dy[i];
            int nx=x+dx[i];
            
            if(ny<0||nx<0||ny>=maxCol||nx>=maxRow){ // map범위 밖
                continue;
            }
            
            if(visit[ny][nx]){ // 방문했으면 
                continue;   
            }
            
            if(maps[ny][nx]==0){ // 벽이면 
                continue;
            }
            
            visit[ny][nx]=visit[y][x]+1;
            q.push({ny,nx});
        }
        
    }
    
    int ans= visit[maxCol-1][maxRow-1];
    
    if(ans==0){
        return -1;
    }
    
    return ans;
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;
class Solution {
    
    int []dx ={-1,1,0,0};
    int []dy ={0,0,-1,1};
    boolean[][]visited = new boolean[101][101];
    
    public int solution(int[][] maps) {
        int answer = 0;
        
        Queue<Pair> q = new LinkedList<>();
        
        q.add(new Pair(0,0,1));
        visited[0][0]=true;
        
        while(!q.isEmpty()){
            
            Pair p = q.poll();
            
            if(p.y==maps.length-1&&p.x==maps[0].length-1)
                return p.cnt;
            
            for(int i=0;i<4;i++){
                
                int ny = p.y+dy[i];
                int nx = p.x+dx[i];
                
                // 행렬 범위를 벗어나면 
                if(ny<0||nx<0||ny>=maps.length||nx>=maps[0].length)
                    continue;
                
                // 좌표가 이미 방문 했으면 
                if(visited[ny][nx])
                    continue;

                // 벽으로 막혀있으면 
                if(maps[ny][nx]==0)
                    continue;      
                
                visited[ny][nx] = true;
                q.add(new Pair(ny,nx,p.cnt+1));
            }
        }
        
        return -1;
    }
}

class Pair{
    
    int y;
    int x;
    int cnt=0;
    Pair(int y,int x,int cnt){
        this.y=y;
        this.x=x;
        this.cnt=cnt;
    }
    
}